ipsec: ensure all trace events are discarded when mon. agg. is enabled

learnitall opened the pull request:
This commit modifies calls to `send_trace_notify` in the datapath to drop trace events related to encrypted packets when monitor aggregation is enabled. More specifically, this commit ensures that whenever `send_trace_notify` is called with a `trace_reason` of `TRACE_REASON_ENCRYPTED`, the `monitor` argument is set to zero. A Coccinelle script is provided in this commit to add a build-time check for this requirement moving forward.

This change helps to reduce the overall CPU usage of Cilium Agents when IPSec encryption is enabled, by reducing the number of trace events emitted by the datapath. Normally monitor aggregation can be used in order to reduce the number of trace events, however IPSec-related trace events are not able to be aggregated since they lack the necessary connection tracking information. See the function `emit_trace_notify` in `bpf/lib/trace.h` for more information.

This endeavor was first started in [cilium/cilium#27168](https://github.com/cilium/cilium/pull/27168), however, in said PR I missed a couple of spots where this type of change could also be applied. This PR acts as a follow-up to ensure that all places where `TRACE_REASON_ENCRYPTED` appears in a call to `send_trace_notify`, the argument `monitor` is set to zero.

Please ensure your pull request adheres to the following guidelines:

- [X] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [X] All code is covered by unit and/or runtime tests where feasible.
- [X] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [X] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [X] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [X] Provide a title or release-note blurb suitable for the release notes.
- [X] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [X] Thanks for contributing!

```release-note
Fix and prevent future bugs limiting pod-to-pod network performance under high load when tunneling and IPSec are both enabled.
```

learnitall committed:
ipsec: ensure all trace events are discarded when mon. agg. is enabled

This commit modifies calls to `send_trace_notify` in the datapath to
drop trace events related to encrypted packets when monitor aggregation
is enabled. More specifically, this commit ensures that whenever
`send_trace_notify` is called with a `trace_reason` of `TRACE_REASON_ENCRYPTED`,
the `monitor` argument is set to zero. A Coccinelle script is provided
in this commit to add a build-time check for this requirement moving forward.

This change helps to reduce the overall CPU usage of Cilium Agents when IPSec
encryption is enabled, by reducing the number of trace events emitted by
the datapath. Normally monitor aggregation can be used in order to
reduce the number of trace events, however IPSec-related trace events
are not able to be aggregated since they lack the necessary connection
tracking information. See the function `emit_trace_notify` in `bpf/lib/trace.h`
for more information.

This same change was applied in `bpf/bpf_lxc.c` in commit 3e52822.

Thank you to Lorenz who added a workaround for passing the verifier with
Clang 10.

See also: cilium/cilium#27168

Co-authored-by: Lorenz Bauer <lmb@isovalent.com>
Signed-off-by: Ryan Drew <ryan.drew@isovalent.com>

pchaigno reviewed (approved):
Nice fix, perfect commit description, and impressive work on Coccinelle! :rocket:

lmb committed:
complexity-tests: add bpf_network configuration

Add an initial bpf_network configuration. The defines are taken from a
sysdump for a failing CI run that seems to exhibit verifier problems.

Signed-off-by: Lorenz Bauer <lmb@isovalent.com>

jrfastab commented on bpf/bpf_network.c:
```diff
@@ -79,10 +87,12 @@ int cil_from_network(struct __ctx_buff *ctx)
 	send_trace_notify(ctx, obs_point_from, 0, 0, 0,
-			  ctx->ingress_ifindex, reason, TRACE_PAYLOAD_LEN);
+			  ingress_ifindex,
+			  trace.reason, trace.monitor);
```
I wouldn't do it as part of this fix, but might be nice if the last arg was just the trace_ctx?

learnitall replied:
I agree, that would be nice. Would you be up for creating an issue? It could probably be tagged as a good first issue.

ldelossa reviewed (approved):
LGTM
